lib/spack/spack/spec.py

```diff
@@ -1559,8 +1559,24 @@ def _set_compiler(self, compiler):
     def _add_dependency(self, spec, deptypes):
         """Called by the parser to add another spec as a dependency."""
         if spec.name in self._dependencies:
-            raise DuplicateDependencyError("Cannot depend on '%s' twice" % spec)
+            # allow redundant compatible dependency specifications
+            # depspec equality checks by name, so we need to check components
+            # separately to test whether the specs are identical
+            orig = self._dependencies[spec.name]
+            for dspec in orig:
+                if deptypes == dspec.deptypes:
+                    try:
+                        dspec.spec.constrain(spec)
+                        return
+                    except spack.error.UnsatisfiableSpecError:
+                        raise DuplicateDependencyError(
+                            "Cannot depend on incompatible specs '%s' and '%s'"
+                            % (dspec.spec, spec)
+                        )
+            else:
+                raise DuplicateDependencyError("Cannot depend on '%s' twice" % spec)
```

Comment:
```suggestion
            orig = self._dependencies[spec.name]

            try:
                dspec = next(dspec in orig if deptypes == dspec.deptypes)
            except StopIteration:
                raise DuplicateDependencyError("Cannot depend on '%s' twice" % spec)

            try:
                dspec.spec.constrain(spec)
            except spack.error.UnsatisfiableSpecError:
                raise DuplicateDependencyError(
                    "Cannot depend on incompatible specs '%s' and '%s'" % (dspec.spec, spec)
                )
```

i still don't like deeply nested control flow, just leaving this as a suggestion.